modules/post/multi/manage/sudo.rb

```diff
@@ -99,7 +99,7 @@ def askpass_sudo(password)
         ::Timeout.timeout(120) do
           vprint_status "Writing the SUDO_ASKPASS script: #{askpass_sh}"
           cmd_exec("echo \\#\\!/bin/sh > #{askpass_sh}") # Cursed csh
-          cmd_exec("echo echo #{password} >> #{askpass_sh}")
+          cmd_exec("echo 'echo '\"'\"'#{password}'\"'\"'' >> #{askpass_sh}")
```

Comment:
This approach will likely also have issues with other shell metacharacters such as ``` ` ```. Backslashes will likely cause issues too.

Given that the whole reason this dodgy code exists is to simply write a file to disk, it would make more sense to use the `write_file` method from the `Msf::Post::File` mixin which was specifically created for this purpose. Internally, this method uses Meterpreter methods if available or `_write_file_unix_shell` for command shell sessions, which takes care of escaping and substitution issues automatically.

The `Msf::Post::File` mixin is already used by this module, which means you can simply do this:

```ruby
          write_file(askpass_sh, "#!/bin/sh\necho '#{password}'\n")
```

This approach has the same flaw with `'` causing issues, but it's much cleaner and easier to read than using 15 levels of escaping using multiple forks with `cmd_exec`, and uses existing libraries for consistency as an added bonus.

Reply:
Agreed, a much cleaner solution. It worked perfectly and I will commit this change.

Reply:
Not sure what happened in your commit. There's a trailing `q` character.

Also, this line should be removed:

```ruby
cmd_exec("echo \\#\\!/bin/sh > #{askpass_sh}") # Cursed csh
```